### DISCUSSION TITLE
Display chapter on an odd page, preceded by a blank even page

### DISCUSSION THREAD

**Discussion by Anh-Kagi** [typst/typst#4337] [Show and tell]
Ok, so I was tweaking with the wonderous-book template. I was trying to have my headings on odd pages, preceded by a blank page.

I was simply doing this with a show rule:
```typst
// SPDX-License-Identifier: MIT
#show heading.where(level: 1): it => {
  pagebreak() // to be on a blank page
  pagebreak(to: "odd", weak: false) // to forcefully insert a blank page before our odd page
  
  it
}
```

But it didn't fit as the page headers were offset: It was difficult to know if the current page contained a heading, since the heading could actually be several pages before the actual chapter title due to the multiple page breaks needed.

The fact that my show rule didn't actually place the page breaks before my heading, but instead inserting them inside the heading was quite annoying and I was wondering if there was a way to actually place them before (if somebody knows, I'd love to hear) but I couldn't find any way.

After quite some digging and some inspiration of #3122, I came with my solution, which you can check here: https://typst.app/project/rq4olhJekGOtRcd41CkM71

Basically, I used a show rule to store the chapters' name and actual page number in a state variable:
```typst
// SPDX-License-Identifier: MIT
#show heading.where(level: 1): it => {
  // jump to odd page
  pagebreak(to: "odd", weak: false)
  context {
    // store actual heading page (after page breaks) in a list
    let current-chapter = (here().page(), it)
    state("chapters").update(st => {
      st.push(current-chapter)
      st
    })
  }
  it
}
```

Then, in my heading template, I retrieve the final list using `state(...).final()`.
I check if there are any chapter in the current page. Otherwise, I get the latest chapter preceding the current page and display its name.
```typst
// SPDX-License-Identifier: MIT
#set page(header: locate(loc => {
  // get all registered chapters (page and name)
  let chapters = state("chapters").final()

  if not chapters.any(it => it.at(0) == loc.page()) { // skip if any chapter in current page
    let last-chapter = chapters.fold((0, none), (acc, it) => {
      if acc.at(0) <= it.at(0) and it.at(0) < loc.page() {
        it
      } else {
        acc
      }
    })
    // force size and weight to keep header homogenous
    text(15pt, weight: "regular", last-chapter.at(1))
  }
}))
```

My solution is certainly not the best, but at least I found one that fit my needs, and I hope it can help other too.

**Comment by jamesrswift** [comment]
The hydra package is probably a good fit here by using it's "book" setting

**Reply by Anh-Kagi** [reply]
I didn't know about hydra.
Unfortunately, it has the same problem with headings being messed up when I force a blank page before my chapter title.

**Comment by Anh-Kagi** [comment]
I've tested another solution that works better for me: I've found a way to insert my page breaks BEFORE my chapters (and therefore not inside them, as in a show rule).

My previous solution had a rather annoying problem: when exporting to pdf, if I clicked on a chapter link, I'd be redirected a few pages before the actual chapter page (because the page breaks are all the same IN the chapters, which therefore appear before the chapter name).

So I looked for a solution to insert my page breaks before my headings: I created a show rule to go through the whole document. I look at the type of all the elements, and when I come across a heading, I add my page breaks and then the heading:
```typst
// SPDX-License-Identifier: MIT
// insert pagebreaks before my headings
#let prepend-headers(it) = {
  let f = it.func()
  if repr(f) == "sequence" { // using repr() because I could not find element function to compare with
    for i in it.children {
      prepend-headers(i)
    }
  } else if repr(f) == "styled" {
    // copy fields style
    let fields = it.fields()
    let child = fields.remove("child")
    let styles = fields.remove("styles")
    f(prepend-headers(child), styles, ..fields)
  } else if f == heading and it.depth == 1 { // search for chapter headings
    // insert page breaks to have chapter on odd page, preceded by a blank even page
    pagebreak()
    pagebreak(to: "odd", weak: false)
    it
  } else {
    it
  }
}
#show: prepend-headers
```
example project: https://typst.app/project/rYdnzqZ5eTESWiZGSt9yH5

It's a bit over-complicated, especially as I don't think I'm taking all cases into account. But it works better for me: my page breaks are now BEFORE my headings. This simplifies the code for adding the chapter title in the header, and the links to the chapters redirect to the chapter page.

This code doesn't go through all the elements though: it doesn't recursively enter all the fields that may contain content to insert page breaks in front of any headings that might be there. I didn't want my code to be too complicated. It just fits in the sequences because I'm assuming that chapter headings won't be contained in atypical elements, but that's modifiable.

This was actually the first solution I thought of, but I found it a bit too complicated to implement right away. If anyone knows of an easier way of inserting my page breaks before my headings, I'd love to hear about it, as my solution is still a little too complicated for my taste.

**Comment by Anh-Kagi** [comment]
Thanks to Typst 0.12.0, I don't need my function anymore 🥳
I can just add those pagebreaks directly in the show rule:
```typst
#show heading: it => pagebreak() + pagebreak(to: "odd", weak: false) + it
```
(Thanks to #4840)

**Comment by jhalmen** [comment]
In case you also want to be able to omit headers/footers/page-numbering on the empty pages, i finally figured out how to do that without drawbacks over in https://github.com/typst/typst/discussions/3122#discussioncomment-11022031
